packages/auth-ts/src/model/additional_user_info.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ProviderId } from '../core/providers';
+import { UserCredential } from './user_credential';
+import { IdTokenResponse, parseIdToken } from './id_token';
+
+export interface AdditionalUserInfo {
+  readonly isNewUser: boolean;
+  readonly profile: { [key: string]: unknown } | null;
+  readonly providerId: ProviderId | null;
+  readonly username: string | null;
+}
+
+export function getAdditionalUserInfo(
+  userCredential: UserCredential
+): AdditionalUserInfo | null {
+  return additionalInfo.get(userCredential) || null;
+}
+
+export function updateAdditionalUserInfoFromIdTokenResponse(
+  userCredential: UserCredential,
+  idTokenResponse: IdTokenResponse
+) {
+  const additionalUserInfo = additionalUserInfoFromIdTokenResponse(
+    idTokenResponse
+  );
+  if (!!additionalUserInfo) {
+    additionalInfo.set(userCredential, additionalUserInfo);
+  }
+}
+
+const additionalInfo = new WeakMap<UserCredential, AdditionalUserInfo>();
+
+function additionalUserInfoFromIdTokenResponse(
+  idTokenResponse: IdTokenResponse
+): AdditionalUserInfo | null {
+  const providerId = idTokenResponse.providerId;
+  const profile =
+    typeof idTokenResponse.rawUserInfo === 'string'
+      ? JSON.parse(idTokenResponse.rawUserInfo)
+      : {};
+  const isNewUser =
+    !!idTokenResponse.isNewUser ||
+    idTokenResponse.kind === 'identitytoolkit#SignupNewUserResponse';
+  if (!providerId && idTokenResponse.idToken !== undefined) {
```

Comment:
(!providerId && idTokenResponse) {   -- no need to check strict inequality with undefined